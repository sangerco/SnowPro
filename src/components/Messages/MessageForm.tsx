import React, { useEffect, useState } from "react";
import { createMessage } from "../../redux/slices/messageSlice";
import { fetchUserData } from "../../redux/slices/userSlice";
import {
  Button,
  DropdownItemProps,
  Form,
  Dropdown,
  Grid,
  Segment,
  Rail,
  Loader,
  Dimmer,
  Card,
  Header,
} from "semantic-ui-react";
import { RootState, AppDispatch } from "../../redux/store";
import { useDispatch, useSelector } from "react-redux";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import { Link, useNavigate } from "react-router-dom";

const MessageForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const senderId = userId ? userId : "";

  const initialMessageState = {
    sender_id: senderId,
    recipient_id: "",
    subject: "",
    body: "",
  };

  console.log(initialMessageState);

  useEffect(() => {
    dispatch(fetchUserData);
  }, [dispatch]);

  const users = useSelector((state: RootState) => state.users.users);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  let userOptions:
    | { key: string; text: string; value: string }[]
    | DropdownItemProps[]
    | undefined;

  if (users) {
    userOptions = users.map((user) => ({
      key: user.id,
      text: user.username,
      value: user.id,
    }));
  } else {
    userOptions = [];
  }
  const [formData, setFormData] = useState(initialMessageState);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(createMessage(formData));

    navigate(-1);

    setFormData(initialMessageState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleBodyChange: React.ChangeEventHandler<HTMLTextAreaElement> = (
    e
  ) => {
    const { name, value } = e.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleDropdownChange = (
    e: React.SyntheticEvent<HTMLElement>,
    data: any
  ) => {
    setFormData({
      ...formData,
      recipient_id: data.value,
    });
  };

  console.log(formData);

  return (
    <div id="background-container">
      <Grid centered columns={3}>
        <Grid.Column>
          <Segment>
            <Form onSubmit={handleSubmit}>
              <label>Who do you want to send this message to?</label>
              <Dropdown
                required
                clearable
                multiple
                options={userOptions}
                fluid
                selection
                onChange={handleDropdownChange}
              />
              <Form.Field>
                <input
                  placeholder="Whatcha talking about?"
                  name="subject"
                  value={formData.subject}
                  onChange={handleChange}
                />
              </Form.Field>
              <Form.TextArea
                placeholder="Message"
                name="body"
                value={formData.body}
                onChange={handleBodyChange}
              />
              <Button type="submit">Send</Button>
            </Form>

            <Rail dividing position={"left"}>
              {skiAreaState.loading ? (
                <Dimmer active>
                  <Loader>Loading...</Loader>
                </Dimmer>
              ) : skiAreas && skiAreas.length > 0 ? (
                <Card>
                  {skiAreas.map((sa) => (
                    <Card.Content key={sa.slug}>
                      <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                    </Card.Content>
                  ))}
                </Card>
              ) : skiAreaState.error ? (
                <Dimmer active>
                  <Header as="h1">
                    Error! Ski Area Data cannot be retrieved!{" "}
                    {skiAreaState.error}
                  </Header>
                </Dimmer>
              ) : null}
            </Rail>

            <Rail dividing position={"right"}></Rail>
          </Segment>
        </Grid.Column>
      </Grid>
    </div>
  );
};

export default MessageForm;
