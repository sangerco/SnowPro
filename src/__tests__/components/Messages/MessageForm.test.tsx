/* eslint-disable testing-library/no-node-access */
import React from "react";
import MessageForm from "../../../components/Messages/MessageForm";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Message Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    message: {
      message: null,
      messages: null,
      sentMessages: null,
      loading: false,
      error: null,
    },
    users: {
      users: [
        {
          userId: "22",
          username: "testuser2",
        },
        {
          userId: "33",
          username: "testuser3",
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <MessageForm />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <MessageForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/new-message")
      .reply(201, {
        message: {
          id: "message-1",
          senderId: "testuser",
          recipientId: "testuser2",
          subject: "test message subject",
          body: "test message body",
          isRead: "false",
          createdAt: "fake created date",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, [
        {
          slug: "vail",
          name: "Vail",
        },
        {
          slug: "arapahoe-basin",
          name: "Arapahoe Basin",
        },
      ]);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <MessageForm />
        </BrowserRouter>
      </Provider>
    );

    expect(pendingMocks().length).toBe(2);

    const subject = screen.getByPlaceholderText("Whatcha talking about?");
    fireEvent.change(subject, { target: { value: "test subject" } });

    const body = screen.getByPlaceholderText("Message");
    fireEvent.change(body, { target: { value: "test message body" } });

    const messageForm = await screen.findByTestId("message-form");
    fireEvent.submit(messageForm);

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
