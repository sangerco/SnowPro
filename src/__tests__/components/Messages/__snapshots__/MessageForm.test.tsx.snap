// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Message Form should match snapshot 1`] = `
<DocumentFragment>
  <div
    id="background-container"
  >
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="four wide column"
        >
          <div
            class="ui segment"
            style="overflow: auto; max-height: 75vh;"
          >
            <div
              class="ui card"
              id="ski-area-card"
            >
               
              <div
                class="content"
              >
                <a
                  href="/ski-areas/ski-area-1"
                >
                  Ski Area 1
                </a>
              </div>
              <div
                class="content"
              >
                <a
                  href="/ski-areas/ski-area-1"
                >
                  Ski Area 1
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="ui raised segment"
          >
            <form
              class="ui form"
              data-testid="message-form"
            >
              <label>
                Who do you want to send this message to?
              </label>
              <div
                aria-expanded="false"
                aria-multiselectable="true"
                class="ui fluid multiple selection dropdown"
                data-testid="select"
                required=""
                role="listbox"
                tabindex="0"
              >
                <i
                  aria-hidden="true"
                  class="dropdown icon"
                />
                <div
                  class="menu transition"
                >
                  <div
                    aria-checked="false"
                    aria-selected="true"
                    class="selected item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      testuser2
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      testuser3
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="field"
              >
                <input
                  name="subject"
                  placeholder="Whatcha talking about?"
                  value=""
                />
              </div>
              <div
                class="field"
              >
                <textarea
                  name="body"
                  placeholder="Message"
                  rows="3"
                />
              </div>
              <button
                class="ui button"
                type="submit"
              >
                Send
              </button>
            </form>
          </div>
        </div>
        <div
          class="four wide column"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
